crates/storage/provider/src/changesets_utils/trie.rs

```diff
@@ -0,0 +1,166 @@
+    }
+
+    fn seek_cursor(&mut self, path: Nibbles) -> Result<(), DatabaseError> {
+        self.cursor_current = self
+            .cursor
+            .seek_by_key_subkey(self.hashed_address, StoredNibblesSubKey(path))?
+            .map(|e| (e.nibbles.0, e.node));
+        Ok(())
+    }
+}
+
+impl<'cursor, P, C> Iterator for StorageTrieCurrentValuesIter<'cursor, P, C>
+where
+    P: Iterator<Item = Nibbles>,
+    C: DbDupCursorRO<tables::StoragesTrie>,
+{
+    type Item = Result<(Nibbles, Option<BranchNodeCompact>), DatabaseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            match (self.paths.peek(), self.cursor_current.as_mut()) {
+                (None, _) => {
+                    // If there are no more paths then there is no further possible output.
+                    return None
+                }
+                (Some(path), None) => {
+                    // If there is a path but the cursor is empty then that path has no node.
+                    let path = *path;
+                    self.paths.next();
+                    return Some(Ok((path, None)))
+                }
+                (Some(path), Some((cursor_path, cursor_node))) => {
+                    // There is both a path and a cursor value, compare their paths.
+                    match path.cmp(cursor_path) {
+                        Ordering::Less => {
+                            // If the path is behind the cursor then there is no value for that
+                            // path, increment the path iter and produce None.
+                            let path = *path;
+                            self.paths.next();
+                            return Some(Ok((path, None)))
+                        }
+                        Ordering::Equal => {
+                            // If the target path and cursor's path match then there is a value for
+                            // that path, increment the path iter and return the value. We don't
+                            // increment the cursor here, that will be handled on the next call to
+                            // `next` in the `Ordering::Greater` branch if the path iter is not
+                            // None.
+                            let cursor_node = core::mem::take(cursor_node);
+                            self.paths.next();
+                            return Some(Ok((*cursor_path, Some(cursor_node))))
+                        }
+                        Ordering::Greater => {
+                            // If the path is ahead of the cursor then seek the cursor forward and
+                            // loop. The cursor will either seek to the path or beyond it.
+                            let path = *path;
+                            if let Err(err) = self.seek_cursor(path) {
+                                return Some(Err(err))
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
does this ever loop more than once? the final branch is the only one that doesn't return, and it looks like as long as `seek_cursor` works as expected, we should return right after that. If so, maybe we can do something like:
```rust
let curr_cursor = self.cursor_current.as_mut();

let Some(curr_path) = self.paths.peek() else {
   return None
};

if curr_cursor.is_some_and(|(cursor_path, _)| curr_path > cursor_path) {
    if let Err(err) = self.seek_cursor(*path) {
        return Some(Err(err))
    }
}

// do big match on curr_cursor